README.md
# The Tiny Tree (DnD Discord Bot)
## Readme

DnD discord helper for the Dungeons and Diseases game. This is a post-FYP project written by myself, and I'll add occassional things to it as and when the inspiration strikes. 

The Order of the Tiny Tree is a fictional religion within my DnD campaign (mostly because a couple of us wanted a ridiculous, out-of-this-world religion) and involves a sect of Tortle monks who grow bonsais on their shells and are pretty extremist eco-terrorists (you can guess where this is going).

Commands for this bot are all called with the `/` character. There are currently the following functions:

1. **_Dice_**, Called with the `/r` prefix. Rolls an `nDm` number of dice - where `n` is the number of dice and `m` is the number of sides per dice. This has several modifiers that can be added:
	- `ndm + ndm - ndm + x - y`, which follows the format of `n` `m`-sided die, and `x` and `y` are both constant modifiers to the roll. `+` and `-` should work anywhere in the statement (including right at the start), but note that any white space is ignored, so if you miss a `+` somewhere the numbers are just mushed together.
	- `advantage` | `adv` |  `a`, followed by the dice roll. All add advantage to your roll (i.e. roll the dice twice). They do not automatically take the higher of the two rolls - since sometimes a crit may mean you have an overall lower score.
	- `disadvantage` | `disadv` |  `d`, followed by the dice. Similarly, all add disadvantage to your roll (i.e. roll the dice twice). It will display both rolls for your viewing pleasure.
	- `initiative` | `init` only works if the DM has started the preparation phase for combat (i.e. the bot will tell you to roll for initiative. Realistically it would be simpler to have rolled initiative for everyone, but where's the fun in that...
	- `/verbose` is a separate command to `/r`, and toggles the amount of information that is shown by the bot. There are some cute prompts to tell you how much information is being shown.
	- Some neat tricks for `/r` include being able to let you know if a nat20 or nat1 has been rolled (only applies if the first dice rolled is a `1d20`).

2. **_Helper_**, a fairly useful cog that gives you information about DnD stuff. Commands are:
	- `/spell` which gives you information about any spell within *Elemental Evil*, *The Player's Handbook*, *Sword Coast Advanced Guide*, and *Xanathar's Guide to Everything*. It follows the format `/spell spell_name`. This is case-insensitive, but you need to have spelled and put and important punctuation in. So, `/spell aGGaNazzAR's SCORcher` will work, but `/spell animalfriendship` will not.
	- `/currency` tells you a bit about the conversion rate between currencies. `/currency info` brings up some useful blurb about money, especially for people (like myself) who can never remember what they all mean. `/currency convert N gp to sp` lets you convert `N` of any currency to another. This is case insensitive, and takes currency arguments as either `gp` or `gold`, `sp` or `silver`, etc.

3. **_Player_** gets you *your own* information - you can't just go mess with someone else's stats or money!
	- `money` brings up the money you currently own. This accesses data from a Google Sheet somewhere on the interwebs, but is pre-cached on the bot in order to speed up responses when you make any requests. If you absolutely **need** to update the money now because you made a change in Google Sheets, you can simply `/reload player` and it should re-download all of the necessary information.
	- `money add n gold` | `money remove n gold` will add or remove any particular currency (not just gold) of a particular amount to your own inventory. This still only works with the specific "gold" or "silver" (etc.) words. This is an immediate Google Sheets request so will take a bit longer than the other commands.
	- `stats` brings up your stats as updated by the DM at some point. If there has been a level up, this will have been done by the DM, you shouldn't need to have to update it yourself - I understand that everyone has their own favourite ways to keep data, so kept away from too much automation.

4. **_Party_** retrieves the party inventory from a .json file. jsons are fast, so I like them (This inventory is only stored locally, and you can't really just mess around with it at will - I'll add Google Sheets integration at some point).
	- `inventory` | `inv` brings up the total party inventory (money and items).
	- `money` shows all of the various currencies that the party jointly holds. You can add/remove money with `/party money add 50 gold`/`/party money remove 50 gold`.
	- `items` brings up the items that the players holds (sans money.) You can add items as well, in the format `/party items add item name; value; item description`. You can use either `[ ; ]` or `[ / ]` as a seperator, and the value should be an integer number in gp (the item name and description can be of any length). Similarly, `/party items remove item name` will remove that item from the party inventory.

5.  **_Combat Helper_**, helps to start initiative, etc. Only rollable by DM. Does a check every 10 seconds to see if it needs to pull data from Google Drive, and if it does it will load all initiative and Pulls data from Google Drive while everyone is rolling initiative. (Within 10 seconds, so don't start combat too soon after initiating it)
	- `/combat start` will start the initiative-rolling phase if the party is currently outside combat, and starts combat if the party is currently in the initiative phase. When combat is started, all encounter creatures have initative rolled for them automatically, and the DM should have already updated the google form *before* starting the preparation phase.
	- `/combat prepare` | `combat prep` moves the party in the initiative-rolling phase. When in this phase, players can use the `/r init` to roll for initiative. `/r init adv` rolls initiative with advantage automatically.
	- `/combat end` ends combat, like it says.
	- `/combat state` will tell you which phase you're currently in. Again, only the DM can do this.

6. **_No Category_** are just some general-purpose commands that help you out a little.
	- `/load Cog` loads a particular module (all the sections above are "Cogs" in `discord.py`, or classes in Python.
	- `/unload Cog` unloads...
	- `/reload Cog` unloads... then loads.
	- `/status` just gives you some nice confirmation that the bot is online by printing some text.


Google Sheets requests run in the background as much as possible using `@tasks.loop` so that commands are responded to faster (particularly initiative rolls, unfortunately money updates still need an immediate update to happen so will take a while).

Last updated: 15 May 2020 (Check the changelog for any newer updates)

---

### Links:

Note: The only bits of this bot that will work for you are the `/helper` and `/dice` Cogs, so I've removed links that let you add the bot to your server. (Everything else is a custom build for my game). Feel free to drop me an e-mail if you'd like some help setting up for your own game!

- [Discord.py Reference](https://discordpy.readthedocs.io/en/latest/index.html#)
- [Github Page (Probably this page)](https://github.com/Joncallim/The-Tiny-Tree)

changelog.md
# The Tiny Tree (Discord DnD Bot)

## Changelog

### 27 May 2020:

- Minor bug fix to do with initiative rolls.
- Beginning to switch bot's output text over from Markdown to Embedded text.
	- Reworked dice roll text to be a little bit prettier, with embedded text and mentions included.
- Fix for the `/party` cog not flagging a non-existent item if a player tries to remove it, also added a flag for adding an item that already exists (it used to just overwrite anything with the same name).
	

- Some rewrite of the `combat_helper` cog:
	- Added a `combat_class` class to help with managing the party during combat, and re-worked some combat initialisation mechanisms to allow for easier expansion in the future.
	- Created the `/kill X` function for the DM, allowing characters to be removed from the turn order.
	- Created the `/turnorder` function, which allows anyone from the party to view the current turn order.

	
- Rewrite of `helper` cog:
	- Removed the `convert` command from `currency`, and gave it a seperate command. (i.e. `/convert 10 gp to sp` now as opposed to `/currency convert 10 gp to sp`.
	- Added a `/currency rates` command to obtain all the exchange rates.
	- Rewrote the spell helper so it's not accessing the spell JSON every time, and pre-loading the spell data to memory on first initialise.
	- Reworked the output text so it's all embedded text now.
	- Reworked the cog so that currency and spells are a seperate function each, making for slightly simpler coding for future versions.


---

### 15 May 2020:

- First release ready for use. Stable release hosted on Google Cloud compute, bot deployed to Dungeons and Diseases. Cogs are:
	1. Combat Helper
	2. Dice
	3. Helper
	4. Party 
	5. Player 
	
- Todo:
	- Introduce money transfer function.
	- Google Docs integration for party inventory (current handling is exclusively in .json locally)
	- Ambient noise generator with a nice big library of looping sounds (rainforest, cave, etc.)
	- Picture commands for DM to introduce images quickly if needed.
	- Individual inventory management.

---


cogs/party.py
#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Created on Thu May 14 13:15:06 2020

@author: Jonathan
"""


import discord
from discord.ext import commands
import re

import json
import string



class party(commands.Cog):
    
    def __init__(self, bot):
        self.add_money = True
        self.bot = bot
    
    def check_inventory(self):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        # For each money item in the "money" dict, prints a single line, then
        # groups them line by line for each currency value.
        moneyString = ""
        for key,value in inventory['money'].items():
            moneyString = "{}{}: [{}]\n".format(moneyString, string.capwords(key), value)
        itemString = ""
        for key,details in inventory['items'].items():
            itemString = "{}\n\n{} (Value: [{} gp])\nDescription: {}".format(itemString, key, details['value'], details['desc'])
        inventoryEmbed = discord.Embed(title="Party Inventory",
                                        description="The party's joint assets, including all money and items:",
                                        color=0xECF0F1)
        inventoryEmbed.add_field(name = "Money:",
                                 value = moneyString,
                                 inline = False)
        inventoryEmbed.add_field(name = "Items:",
                                 value = itemString,
                                 inline = False)
        return inventoryEmbed
    
    
    def check_items(self):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        itemString = ""
        for key,details in inventory['items'].items():
            itemString = "{}\n\n{} (Value: [{} gp])\nDescription: {}".format(itemString, key, details['value'], details['desc'])
        inventoryEmbed = discord.Embed(title="Party Items",
                                        description="The party's jointly-owned items:{}".format(itemString),
                                        color=0xECF0F1)
        return inventoryEmbed
    
    def add_item(self, new_item):
        # There are two splitters possible - ; and / are both valid characters.
        # This takes the first entry as the name, the second entry as the item's
        # value in gp, and the third as the description. Joins all the items
        # with a single space now, but still splits them using the symbols.
        item_details = re.split('\;|\/',' '.join(new_item))
        # Takes the first entry as the item's name - capitalises the first letter
        # of every word, so 'horses of gondor' becomes 'Horses Of Gondor" which
        # isn't ideal but hey ho...
        item_name = string.capwords(item_details[0])
        # Accesses the json file with the party inventory - json used as choice
        # of database, since jsons are FAST.
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        if item_name in inventory['items']:
            inventoryEmbed = discord.Embed(title="Item already in inventory!",
                                           description="Use `/party items` to check what the party currently owns.",
                                           color=0xECF0F1)
            return inventoryEmbed
        else:
            # Second item should be the value. If a number hasn't been entered, 
            # someone has messed it up and it won't be registered. If no entry, the
            # value is set to 0. Also handles description similarly, but capitalises
            # the first word of every sentence here.
            if len(item_details) == 1:
                item_value  = 0
                item_description = "No Description"
            elif len(item_details) == 2:
                print('here')
                item_value = int(item_details[1].strip())
                item_description = "No Description"
            else:
                item_value = int(item_details[1])
                item_description = item_details[2].strip().capitalize()
            # creates a new dictionary with the item name specified, and then updates
            # the overall inventory dictionary to include this new item dictionary.
            newItem = {item_name : {'desc': item_description, 'value': item_value}}
            inventory['items'].update(newItem)
            # Saves the new inventory to the json file. sort_keys and indent are 
            # used to beautify the file, making for easier human-reading.
            with open('files/party_inventory.json', 'w') as f:
                 f.write(json.dumps(inventory, sort_keys=True, indent=4))
            inventoryEmbed = discord.Embed(title="Item Added!",
                                           description="{} has been added to the party inventory.".format(item_name),
                                           color=0xECF0F1)
            inventoryEmbed.add_field(name = "Value:",
                                     value = "{} gp".format(item_value),
                                     inline = False)
            inventoryEmbed.add_field(name = "Description",
                                     value = item_description,
                                     inline = False)
            return inventoryEmbed
    
    def remove_item(self, name):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        
        name = string.capwords(name)
        # Sanity check to make sure the item exists in the inventory. If it 
        # does, then pops it and tells you that it's been removed. If it doesn't 
        # exist, throws an error up.
        if name in inventory['items']:
            inventory['items'].pop(name)
            with open('files/party_inventory.json', 'w') as f:
                 f.write(json.dumps(inventory, sort_keys=True, indent=4))
            inventoryEmbed = discord.Embed(title="Item Removed!",
                                           description="{} has been removed from the party inventory.".format(name),
                                           color=0xECF0F1)
            return inventoryEmbed
        else:
            inventoryEmbed = discord.Embed(title="Item does not exist!",
                                           description="Use `/party items` to check what the party currently owns.",
                                           color=0xECF0F1)
            return inventoryEmbed
    
    def check_money(self):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        # Fairly strightforward - iterates through each currency stored inside
        # the list, and then copies it out to a nice long string for printing.
        moneyString = ""
        for key,value in inventory['money'].items():
            moneyString = "{}{}: [{}]\n".format(moneyString, string.capwords(key), value)
        inventoryEmbed = discord.Embed(title="Party Funds",
                                       description="The party's accountant holds:\n{}".format(moneyString),
                                       color=0xECF0F1)
        return inventoryEmbed
    
    def update_money(self, value, currency):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        # Get old value of currency stored 
        oldValue = inventory['money'].get(currency)
        if self.add_money:
            newValue = oldValue + value
        else:
            newValue = oldValue - value
            # Resets the add_money counter, so the next iteration will be an 
            # addition, if called again.
            self.add_money = True
        # Sanity check. The party can't have a negative amount of money, so if
        # this is the case, returns the statement telling you that you can't 
        # remove that much money!
        if newValue < 0:
            inventoryEmbed = discord.Embed(title="Insufficient Money!",
                                           description="The party does not have that much {}. Type /party money to check how much you currently have.".format(currency),
                                           color=0xECF0F1)
            return inventoryEmbed
        else:
            # Creates a new dict with the correct currency and new value to be 
            # updated, then updates the original list, and pushes it to the json
            # file to save. Again sort_keys and indent are used for beautification
            newMoney = {currency: newValue}
            inventory['money'].update(newMoney)
            with open('files/party_inventory.json', 'w') as f:
                 f.write(json.dumps(inventory, sort_keys=True, indent=4))
            # Returns string to show updated money amounts. Capwords just makes
            # sure that each word is initial-capsed so it looks pretty.
            moneyString = ""
            for key,amount in inventory['money'].items():
                moneyString = "{}{}: [{}]\n".format(moneyString, string.capwords(key), amount)
            inventoryEmbed = discord.Embed(title="Party funds have been updated!",
                                           description="{} {} has been added to the party's funds.".format(value, currency),
                                           color=0xECF0F1)
            inventoryEmbed.add_field(name = "Updated party fund:",
                                     value = moneyString,
                                     inline = False)
            return inventoryEmbed
        
    def remove_money(self):
        with open('files/party_inventory.json', 'r') as file:
            inventory = json.load(file)
        # This is easy, just automatically writes a new dictionary with 
        # everything set to 0, so don't need to iterate through every value
        # in the dictionary - we know exactly what should be in it, and no
        # other currencies.
        newMoney = {'platinum': 0, 'gold':0, 'electrum': 0, 'silver': 0, 'copper': 0}
        inventory['money'].update(newMoney)
        # Pushing new dict to json
        with open('files/party_inventory.json', 'w') as f:
             f.write(json.dumps(inventory, sort_keys=True, indent=4))
        inventoryEmbed = discord.Embed(title="The party now holds no collective funds.",
                                       description="Party funds have been cleared.",
                                       color=0xECF0F1)
        return inventoryEmbed
    
    @commands.Cog.listener()
    async def on_ready(self):
        print('The party Cog is now online.')
    
    @commands.command(name='party', help="Party Inventory Management.\n1. [inventory] will show a list of all items jointly held.\n2. [items] will show only items, and exclude money. You can also write '/party items add new_item value_in_gp description' to add an item, and '/party items remove item_name' to remove any item.\n3. [money] displays the current held cash in various currencies. '/party money add amount currency' will add that amount to the appropriate bin, and '/party money remove amount currency' will remove it. You can also enter '/party money clear' to clean out the entire party's shared money.")
    async def party_helper(self, ctx, *arg):
        # Check for first argument == Inventory/inv
        if (arg[0].lower() == 'inventory') | (arg[0].lower() == 'inv'):
            inventoryEmbed = self.check_inventory()
        elif arg[0].lower()== 'items':
            if len(arg) == 1:
                inventoryEmbed = self.check_items()
            elif arg[1].lower() == 'add':
                # Sends all the arguments through to add_item... Anything else
                # can be handled there quite easily.
                inventoryEmbed =  self.add_item(arg[2:])
            elif arg[1].lower() == 'remove':
                # If you're removing an item from the inventory, just joining
                # all the strings together from the arg will be the item's 
                # name. This  must match the saved text, although should be
                # case-insensitive given how this function handles the text.
                inventoryEmbed =  self.remove_item(' '.join(arg[2:]))
        # This is the section to check if money is being used... Same process
        # as with the inventory.
        elif arg[0].lower() == 'money':
            if len(arg) == 1:
                inventoryEmbed = self.check_money()
            elif arg[1].lower() == 'add':
                # "add 50 gold" would be an appropriate input - the second argument
                # should be a number, so is converted to an integer as such.
                inventoryEmbed =  self.update_money(int(arg[2]), arg[3])
            elif arg[1].lower() == 'remove':
                # Literally the same as above, but add_money check is set to 
                # False so the command knows to remove money and not add to it.
                self.add_money = False
                inventoryEmbed =  self.update_money(int(arg[2]), arg[3])
            # This is the scrub-inventory command, basically clears all the cash
            # out of the inventory so you can distribute funds/use them.
            elif arg[1].lower() == 'clear':
                inventoryEmbed = self.remove_money()
        else:
            inventoryEmbed = discord.Embed(title="Invalid Syntax!",
                                        description="Use `/party money [options]`, where `[options]` must be in the order: `add/remove amount currency`.",
                                        color=0xECF0F1)        
        await ctx.send(embed = inventoryEmbed)
        
def setup(bot):
    bot.add_cog(party(bot))





